usvm-ts/src/main/kotlin/org/usvm/machine/types/FakeExprUtil.kt

```diff
@@ -15,55 +15,52 @@ import org.usvm.machine.interpreter.TsStepScope
 import org.usvm.machine.state.TsState
 import org.usvm.memory.ULValue
 
-fun TsContext.mkFakeValue(
-    scope: TsStepScope,
+fun TsState.mkFakeValue(
     boolValue: UBoolExpr? = null,
     fpValue: UExpr<KFp64Sort>? = null,
     refValue: UHeapRef? = null,
-): UConcreteHeapRef {
+): UConcreteHeapRef = with(ctx) {
     require(boolValue != null || fpValue != null || refValue != null) {
         "Fake object should contain at least one value"
     }
 
-    return scope.calcOnState {
-        val fakeValueRef = createFakeObjectRef()
-        val address = fakeValueRef.address
-
-        val boolTypeExpr = trueExpr
-            .takeIf { boolValue != null && fpValue == null && refValue == null }
-            ?: makeSymbolicPrimitive(boolSort)
-        val fpTypeExpr = trueExpr
-            .takeIf { boolValue == null && fpValue != null && refValue == null }
-            ?: makeSymbolicPrimitive(boolSort)
-        val refTypeExpr = trueExpr
-            .takeIf { boolValue == null && fpValue == null && refValue != null }
-            ?: makeSymbolicPrimitive(boolSort)
-
-        val type = EtsFakeType(
-            boolTypeExpr = boolTypeExpr,
-            fpTypeExpr = fpTypeExpr,
-            refTypeExpr = refTypeExpr,
-        )
-        memory.types.allocate(address, type)
-        scope.assert(type.mkExactlyOneTypeConstraint(ctx))
-
-        if (boolValue != null) {
-            val boolLValue = ctx.getIntermediateBoolLValue(address)
-            memory.write(boolLValue, boolValue, guard = ctx.trueExpr)
-        }
+    val fakeValueRef = createFakeObjectRef()
+    val address = fakeValueRef.address
+
+    val boolTypeExpr = trueExpr
+        .takeIf { boolValue != null && fpValue == null && refValue == null }
+        ?: makeSymbolicPrimitive(boolSort)
+    val fpTypeExpr = trueExpr
+        .takeIf { boolValue == null && fpValue != null && refValue == null }
+        ?: makeSymbolicPrimitive(boolSort)
+    val refTypeExpr = trueExpr
+        .takeIf { boolValue == null && fpValue == null && refValue != null }
+        ?: makeSymbolicPrimitive(boolSort)
+
+    val type = EtsFakeType(
+        boolTypeExpr = boolTypeExpr,
+        fpTypeExpr = fpTypeExpr,
+        refTypeExpr = refTypeExpr,
+    )
+    memory.types.allocate(address, type)
+    pathConstraints += type.mkExactlyOneTypeConstraint(ctx)
```

Comment:
This is not a simple change. Because of it later you can have inconsistent model somewhere. 

You have states: a -> b -> c, and in a you added this path constraint. But it doesn't recalculate models, therefore, when in C you asks info about this type constraint, it might be incorrect since you didn't call solver. Previously, there was an assert that forces this models reevaluation

Reply:
That was your suggestion. How to re-organize all this without an access to the scope?